utils/data-migration/migrate-acbs-utilisation-data-to-tfm/index.ts

```diff
@@ -0,0 +1,61 @@
+import { SqlDbDataSource } from '@ukef/dtfs2-common/sql-db-connection';
+import { FacilityUtilisationDataEntity, ReportPeriodPartialEntity } from '@ukef/dtfs2-common';
+import { DataSource } from 'typeorm';
+import { existsSync, readFileSync } from 'fs';
+import { resolve } from 'path';
+
+const JSON_FILE_PATH = resolve(__dirname, 'acbs-utilisation-data.json');
+
+const REPORT_PERIOD: ReportPeriodPartialEntity = {
+  start: { month: 1, year: 2020 },
+  end: { month: 3, year: 2020 },
+};
+
+type UtilisationDataEntry = {
+  facilityId: string;
+  utilisation: number;
+  fixedFeePremium: number;
+};
+
+const validateReportPeriod = (reportPeriod: ReportPeriodPartialEntity): void => {
+  const { start, end } = reportPeriod;
+  if (start.year > end.year || (start.year === end.year && start.month > end.month)) {
+    throw new Error('Report period end date cannot be before start date');
+  }
+};
+
+const insertFacilityUtilisationEntry = async (dataSource: DataSource, { facilityId, utilisation, fixedFeePremium }: UtilisationDataEntry) => {
+  const facilityUtilisationData = new FacilityUtilisationDataEntity();
```

Comment:
* Singleton pattern is not widely used in DTFS code base, is there a reason this pattern has been used.
* Could you also please refer to Codacy SCA alerts on unsafe member access please.

Reply:
I don't think this is a singleton pattern - we're instantiating a new FacilityUtilisationDataEntity every time we run this function rather than re-using a single entity.

Codacy errors on the readme look normal - on the typescript codacy has been broken for a long time - this FacilityUtilisationDataEntity has a type and is well typed and isn't any - but Codacy seems unable to pick this up properly and so complains. Think this is a wider problem with the Codacy setup unfortunately :(

Reply:
* @ChristianMcc just noticed this pattern is already in use was implemented last month, if we are instantiating a class object I would have expected to use `setter` methods to update the individual properties perhaps we can have a chat about this approach so I understand the context.

* Some of these codacy errors are easily resolavable link to codacy suggestions https://app.codacy.com/gh/UK-Export-Finance/dtfs2/pull-requests/3405, of!course please ignore `import` related errors as they are genuine issue with codacy unable to refer to the deps. Happy to jump on a call regarding this as well if it helps.

Reply:
This is not so much a pattern choice with our entity classes as much as it is something specific to typeorm - we can't start defining `set` and `get` methods as that will start to conflict with the way typeorm works (for example, you can't have `private` properties as the repository API needs to be able to access them). As a compromise, I think we should add a static `create` method to the `FacilityUtilisationData` class which constructs this entity with all the supplied parameters (see `FeeRecordEntity.create` for an example). This also gives us better future proofing if we ever add a property to the table and forget to update this script

Reply:
Added the create method Josh referenced as I quite like that pattern and it's on the other entities already, thanks!